plugins/animate/helix.hpp

```diff
@@ -0,0 +1,329 @@
+#include <wayfire/window-manager.hpp>
+#include <wayfire/view-transform.hpp>
+#include <wayfire/txn/transaction-manager.hpp>
+#include <wayfire/render-manager.hpp>
+#include <wayfire/scene-render.hpp>
+#include <wayfire/util/duration.hpp>
+#include <glm/gtc/matrix_transform.hpp>
+
+#include "animate.hpp"
+
+
+wf::option_wrapper_t<wf::animation_description_t> helix_duration{"animate/helix_duration"};
+wf::option_wrapper_t<int> helix_strip_height{"animate/helix_strip_height"};
+wf::option_wrapper_t<int> helix_rotations{"animate/helix_rotations"};
+
+static const char *helix_vert_source =
+    R"(
+#version 100
+
+attribute mediump vec3 position;
+attribute mediump vec2 uv_in;
+
+uniform mat4 matrix;
+
+varying highp vec2 uv;
+
+void main() {
+    uv = uv_in;
+    gl_Position = matrix * vec4(position, 1.0);
+}
+)";
+
+static const char *helix_frag_source =
+    R"(
+#version 100
+@builtin_ext@
+@builtin@
+
+precision mediump float;
+
+varying highp vec2 uv;
+
+void main()
+{
+    gl_FragColor = get_pixel(uv);
+}
+)";
+
+namespace wf
+{
+namespace helix
+{
+static std::string helix_transformer_name = "animation-helix";
+using namespace wf::scene;
+using namespace wf::animation;
+class helix_animation_t : public duration_t
+{
+  public:
+    using duration_t::duration_t;
+};
+class helix_transformer : public wf::scene::view_2d_transformer_t
+{
+  public:
+    OpenGL::program_t program;
+    wf::geometry_t animation_geometry;
+    helix_animation_t progression{helix_duration};
+
+    class simple_node_render_instance_t : public wf::scene::transformer_render_instance_t<helix_transformer>
+    {
+        wf::signal::connection_t<node_damage_signal> on_node_damaged =
+            [=] (node_damage_signal *ev)
+        {
+            push_to_parent(ev->region);
+        };
+
+        wayfire_view view;
+        damage_callback push_to_parent;
+
+      public:
+        simple_node_render_instance_t(helix_transformer *self, damage_callback push_damage,
+            wf::output_t *output) : wf::scene::transformer_render_instance_t<helix_transformer>(self,
+                push_damage,
+                output)
+        {
+            this->push_to_parent = push_damage;
+            self->connect(&on_node_damaged);
+        }
+
+        ~simple_node_render_instance_t()
+        {}
+
+        void schedule_instructions(
+            std::vector<render_instruction_t>& instructions,
+            const wf::render_target_t& target, wf::region_t& damage) override
+        {
+            instructions.push_back(render_instruction_t{
+                        .instance = this,
+                        .target   = target,
+                        .damage   = damage & self->animation_geometry,
+                    });
+        }
+
+        void transform_damage_region(wf::region_t& damage) override
+        {
+            damage |= wf::region_t{self->animation_geometry};
+        }
+
+        void render(const wf::render_target_t& target,
```

Comment:
Looking at this, I have an idea on how to make this less CPU-intensive, because I am getting 50%+ cpu usage even on 1080p (the effect will be much more pronounced on 4k displays since the algorithm is linear in height):

Compute a vertex buffer with strips spanning a simple [0,1]x[0,1] rectangle (the rectangle can be easily scaled to match the view size with a matrix multiplication in the vertex shader).

Based on the actual view height at runtime, we select how many strips to render.

Rotation is also doable in the vertex shader - for example by using something like `floor(gl_VertexID / 6.0` you can get the strip index, and from the strip index you can calculate the rotation for the strip. 

Does that sound about right? Maybe I misunderstand the implementation, but feels like the computation-intensive part of this effect can thus be offloaded completely to the GPU.